### PR TITLE
CDAP-8478 Catch exceptions thrown in the listener.

### PR DESCRIPTION
**Pull request by sagarkapare**
Catch exceptions thrown by listeners attached to the ProgramController, so that all listeners get opportunity to execute.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java [hunk 1]
```diff
@@ -416,94 +416,104 @@ private ListenerCaller(Listener listener, Executor executor) {
       this.executor = executor;
     }
 
+    private void execute(Runnable runnable, String methodName) {
+      try {
+        executor.execute(runnable);
+      } catch (Throwable t) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java:422]: This only works if the executor is a same thread executor. I think it is better do it in the MultiListenerCaller methods

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java:422]: It is clearer for the intention

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java:422]: Sure. Will move it to `MultiListenerCaller`. However we will still need to handle it in the `addListener` method which directly makes `caller.init` method call bypassing `MultiListenerCaller`, right?

**Comment by sagarkapare** [comment]
@chtyim Addressed comments. Please take a look again. Testing on the cluster with new changes.

**Comment by chtyim** [comment]
Sorry @sagarkapare , I looked at the class again and it seems like your original change is actually better. Can you just revert to that one and merge?

**Comment by sagarkapare** [comment]
Thank you @chtyim for the review. reverted the last commit. build is running here - https://builds.cask.co/browse/CDAP-RUT687-3

**Comment by sagarkapare** [comment]
Build is green here - https://builds.cask.co/browse/CDAP-RUT687-JOB1-5. Merging.

**Pull request merged**